make/common/JavaCompilation.gmk

```diff
@@ -437,26 +438,26 @@ define SetupJavaCompilationBody
       $1_API_DIGEST_FLAGS := \
           -classpath $$(BUILDTOOLS_OUTPUTDIR)/depend \
           -Xplugin:"depend $$($1_API_TARGET)" \
+          "-XDinternalAPIPath=$$($1_API_INTERNAL)" \
           #
 
       $1_EXTRA_DEPS := $$(BUILDTOOLS_OUTPUTDIR)/depend/_the.COMPILE_DEPEND_batch
     endif
 
-    # Create a file with all sources, to pass to javac in an @file.
+    # Do the actual compilation
+    # Creates a file with all sources, to pass it to javac in an @file.
     # $$($1_VARDEPS_FILE) is used as dependency to track changes in set of
     # list of files.
-    $$($1_FILELIST): $$($1_SRCS) $$($1_VARDEPS_FILE)
-		$$(call MakeDir, $$(@D))
-		$$(call LogWarn, Compiling $$(words $$($1_SRCS)) files for $1)
-		$$(eval $$(call ListPathsSafely, $1_SRCS, $$($1_FILELIST)))
-
-    # Do the actual compilation
-    $$($1_COMPILE_TARGET): $$($1_SRCS) $$($1_FILELIST) $$($1_DEPENDS) \
+    $$($1_COMPILE_TARGET): $$($1_SRCS) $$($1_DEPENDS) \
         $$($1_VARDEPS_FILE) $$($1_EXTRA_DEPS) $$($1_JAVAC_SERVER_CONFIG)
 		$$(call MakeDir, $$(@D))
+		$$(call LogWarn, Compiling up to $$(words $$($1_SRCS)) files for $1)
+		$$(eval $$(call ListPathsSafely, $1_SRCS, $$($1_FILELIST)))
+		$$(call WriteFile, "-XDmodifiedInputs=$$? ", $$@.modfiles)
```

Comment:
I'm not sure WriteFile is safe to use here. `$?` could potentially contain all files in `$1_SRCS`, which is why we use ListPathsSafely in the first place. This is only a problem if using make <4.0, but as long as that's not prohibited, we need to keep it working without the file function.

Reply:
Is there some recommendation on how to generate what is needed? Which is:
```
"-XDmodifiedFiles=<list-of-files> "
```

I was looking at the `WriteFile` rule, and it is not clear to be what is the problem there, but I'm not very knowledgeable of the build system.

I was trying to use `ListPathsSafely`, but wasn't able to find a way that would be working.

Reply:
Before make 4.0, there was no $(file ...) function in make. That meant that if you needed to write something to a file directly from the makefile, you had to do something like $(shell echo "foo" > /my/file). This works ok as long as the amount of text you need to write is small enough, but when writing long lists of filenames, you will eventually hit OS limits on command line length (which are different on different OSes, most notably much shorter on Windows). Our solution to this was ListPathsSafely, which breaks up long lists of path names into manageable chunks (and also compresses parts of the path using sed).

In make 4.0, the $(file) function was introduced, which makes this so much easier. Since it was also more performant, I implemented a 4.0 version of ListPathsSafely. However, we haven't changed the build to require make >4.0, so we have to keep both implementations around.

The WriteFile macro was only intended writing smaller files, so doesn't have any of the command splitting functionality of ListPathsSafely. 

So to answer your question, no there isn't a good way to achieve what you are asking for today in a safe way, not without modifying or extending the macros. One could imagine adding a flag to ListPathsSafely that made it use space instead of newline as separator. Another solution could be to have Depend.java accept something like 
```
"-XDmodifiedFilesFile=/path/to/file"
```
where the file is a list as ListPathsSafely would create it.

We could also consider requiring make 4.0, but that will need some socializing first.

Reply:
I can change `Depend` for this, that is easy, but it is not clear how to write the file. I've tried the naive `$$(eval $$(call ListPathsSafely, $$?, $$($1_MODFILES)))` at the end of `$$($1_FILELIST):` but that does not seem to write the modified files into the file - it produces an empty file. Is there a way for me to write the modified files into the file? Thanks.

Reply:
Oh right, ListPathsSafely takes the data to write by reference, not by value (as in you supply the name of a variable that contains the data, not the data itself). I doubt `?` can be sent into an eval as a variable name, but I may be wrong. Perhaps you could extract the value into another value first in a separate line in the recipe, something like this:

```
        $$(eval $1_MODFILES := $$?)
        $$(eval $$(call ListPathsSafely, $1_MODFILES, $$($1_MODFILELIST)))
```
(note that I redefined $1_MODFILES to be the list of modified files and introduced a new variable $1_MODFILELIST for the filename where we print it)

Reply:
> Oh right, ListPathsSafely takes the data to write by reference, not by value (as in you supply the name of a variable that contains the data, not the data itself). I doubt `?` can be sent into an eval as a variable name, but I may be wrong. Perhaps you could extract the value into another variable first in a separate line in the recipe, something like this:
> 
> ```
>         $$(eval $1_MODFILES := $$?)
>         $$(eval $$(call ListPathsSafely, $1_MODFILES, $$($1_MODFILELIST)))
> ```
> 
> (note that I redefined $1_MODFILES to be the list of modified files and introduced a new variable $1_MODFILELIST for the filename where we print it)

Thanks! Done in:
https://github.com/openjdk/jdk/pull/10104/commits/23e3b75679908d26b76cc291a8efba06879c0d83